proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,329 @@
+```json
+{
+  "reactNativeManifest": {
+    "capabilities": {
+      "codegenConfig": {
+        "name": "MyLib",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.example.mypackage"
+        }
+      }
+    }
+  }
+}
+```
+
+### Hermes Support
+
+Similarly to New Architecture support, the current way to enable/disable the Hermes engine is toggled by using a Gradle Property `hermesEnabled` on Android and changing the `:hermes_enabled` property in the Podfile.
+
+We propose to add the `hermes` section under the `reactNativeManifest.capabilities` section as follows:
+
+```json
+{
+  "reactNativeManifest": {
+    "capabilities": {
+      "hermes": {
+        "enabled": true
+      }
+    }
+  }
+}
+```
+
+which will also allow for split configuration between platforms as follows:
+
+```json
+{
+  "reactNativeManifest": {
+    "android": {
+      "capabilities": {
+        "hermes": {
+          "enabled": true
+        }
+      }
+    },
+    "ios": {
+      "capabilities": {
+        "hermes": {
+          "enabled": true
+        }
+      }
+    }
+  }
+}
+```
+
+### React Native Release feature flagging
+
+We currently have a numer of different files where capabilities of React Native can be toggled to enable/disable features in the runtime.
+For example we have:
+* [android/gradle.properties](https://github.com/facebook/react-native/blob/main/packages/react-native/template/android/gradle.properties) to specify key-value build time properties for Android
+* [android/app/build.gradle](https://github.com/facebook/react-native/blob/main/packages/react-native/template/android/app/build.gradle) to specify build time configuration for Android
+* [ReactNativeFeatureFlags.js](https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/ReactNative/ReactNativeFeatureFlags.js) which contains runtime feature toggles for the Javascript layer.
+* [ReactFeatureFlags.java](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/config/ReactFeatureFlags.java) which contains runtime feature toggles for Android.
+* [ReactNativeConfig.cpp](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactCommon/react/config/ReactNativeConfig.cpp) which contains runtime feature toggles for the C++ layer.
+
+Most of those config flags are undocumented and they're not verified for invalid configuration. Specifically, some feature flags require to be toggled on multiple layers, and it's easy to forget to toggle one of them. Invalid combinations of those flags could lead to unexpected runtime behavior.
+
+We propose to define `reactNativeManifest` as the **preferred entry point** for all the user facing entry points. All the tooling should be adapted to consume the information from the `reactNativeManifest` section and have sensible defaults defined by the tooling itself.
```

Comment:
Is this only for apps? I'm guessing it doesn't make sense for a library to specify flags? But I guess there could be a library that enforces a certain feature flag?

Reply:
It's for both apps and libraries. There could be a scenarios where a capability needs all the dependencies to expose such capability (for example the `newArch` one).

Reply:
Would this proposal enforce this? Like if a library specifies a certain feature flag, and an app uses that library, how do we communicate this?

Reply:
> Would this proposal enforce this? Like if a library specifies a certain feature flag, and an app uses that library, how do we communicate this?

That's up to each feature to define the semantic of such flags. That's also why we need a dedicated RFC for each feature. There might be features that have a hard requirement on dependencies (like the New Architecture Renderer) and other features that have fallbacks